File: CS/Form1.cs
using Create_a_Report_Bound_to_XPObjectSource.nwind;
using DevExpress.Xpo;
using DevExpress.Xpo.Metadata;
using DevExpress.XtraReports.UI;
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Configuration;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Forms;

namespace Create_a_Report_Bound_to_XPObjectSource {
    public partial class Form1 : Form {

        // A data layer instance that is used in the ResolveSession/DismissSession event handlers
        IDataLayer DataLayerInstance = null;

        public Form1() {
            InitializeComponent();
        }

        private XPObjectSource CreateDataSource() {
			XPObjectSource dataSource = new XPObjectSource() {
				ConnectionStringName = "nwind"
			};
			dataSource.SetEntityType(typeof(Products));
			return dataSource;
        }

        private XtraReport CreateReport() {
            XtraReport report = new XtraReport();
            DetailBand DetailBand = new DetailBand();
            DetailBand.HeightF = 50;

            XRLabel XRLabel = new XRLabel();
            XRLabel.WidthF = 300;
            XRLabel.ExpressionBindings.Add(new ExpressionBinding("BeforePrint", "Text", "[ProductName]"));

            DetailBand.Controls.Add(XRLabel);
            report.Bands.Add(DetailBand);
            report.DataSource = CreateDataSource();
            return report;
        }

        private void simpleButton1_Click(object sender, EventArgs e) {
            // Show the report's Print Preview
            XtraReport1 report = new XtraReport1();
            ReportPrintTool printTool = new ReportPrintTool(report);
            printTool.ShowRibbonPreviewDialog();
        }

        private void simpleButton2_Click(object sender, EventArgs e) {
            // Open the report in an End-User Report Designer
            XtraReport1 report = new XtraReport1();
            ReportDesignTool designTool = new ReportDesignTool(report);
            designTool.ShowRibbonDesignerDialog();
        }

        private void simpleButton3_Click(object sender, EventArgs e) {
            XtraReport report = CreateReport();
            ReportPrintTool printTool = new ReportPrintTool(report);
            printTool.ShowPreviewDialog();
        }

        private void simpleButton4_Click(object sender, EventArgs e) {
            XtraReport1 report = new XtraReport1();

            XPObjectSource dataSource = (XPObjectSource)report.DataSource;
            dataSource.ResolveSession += new EventHandler<ResolveSessionEventArgs>(OnResolveSession);
            dataSource.DismissSession += new EventHandler<ResolveSessionEventArgs>(OnDismissSession);

            ReportPrintTool printTool = new ReportPrintTool(report);
            printTool.ShowPreviewDialog();
        }

        private void OnResolveSession(object sender, ResolveSessionEventArgs e) {
            //Сreate a single instance of IDataLayer if it is not created
            if (DataLayerInstance == null) {
                string connectionString = ConfigurationManager.ConnectionStrings["nwind"].ConnectionString;
                DataLayerInstance = XpoDefault.GetDataLayer(connectionString, DevExpress.Xpo.DB.AutoCreateOption.SchemaAlreadyExists);
            }

            //Create new session based on the instance of DataLayer
            e.Session = new UnitOfWork(DataLayerInstance);
        }

        private void OnDismissSession(object sender, ResolveSessionEventArgs e) {
            e.Session.Session.Dispose();
        }
    }
}

File: VB/Form1.vb
Imports Create_a_Report_Bound_to_XPObjectSource.nwind
Imports DevExpress.Xpo
Imports DevExpress.Xpo.Metadata
Imports DevExpress.XtraReports.UI
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Configuration
Imports System.Data
Imports System.Drawing
Imports System.Linq
Imports System.Text
Imports System.Threading.Tasks
Imports System.Windows.Forms

Namespace Create_a_Report_Bound_to_XPObjectSource
	Partial Public Class Form1
		Inherits Form

		' A data layer instance that is used in the ResolveSession/DismissSession event handlers
		Private DataLayerInstance As IDataLayer = Nothing

		Public Sub New()
			InitializeComponent()
		End Sub

		Private Function CreateDataSource() As XPObjectSource
			Dim dataSource As New XPObjectSource() With {.ConnectionStringName = "nwind"}
			dataSource.SetEntityType(GetType(Products))
			Return dataSource
		End Function

		Private Function CreateReport() As XtraReport
			Dim report As New XtraReport()
			Dim DetailBand As New DetailBand()
			DetailBand.HeightF = 50

			Dim XRLabel As New XRLabel()
			XRLabel.WidthF = 300
			XRLabel.ExpressionBindings.Add(New ExpressionBinding("BeforePrint", "Text", "[ProductName]"))

			DetailBand.Controls.Add(XRLabel)
			report.Bands.Add(DetailBand)
			report.DataSource = CreateDataSource()
			Return report
		End Function

		Private Sub simpleButton1_Click(ByVal sender As Object, ByVal e As EventArgs) Handles simpleButton1.Click, Me.Click
			' Show the report's Print Preview
			Dim report As New XtraReport1()
			Dim printTool As New ReportPrintTool(report)
			printTool.ShowRibbonPreviewDialog()
		End Sub

		Private Sub simpleButton2_Click(ByVal sender As Object, ByVal e As EventArgs) Handles simpleButton2.Click
			' Open the report in an End-User Report Designer
			Dim report As New XtraReport1()
			Dim designTool As New ReportDesignTool(report)
			designTool.ShowRibbonDesignerDialog()
		End Sub

		Private Sub simpleButton3_Click(ByVal sender As Object, ByVal e As EventArgs) Handles simpleButton3.Click
			Dim report As XtraReport = CreateReport()
			Dim printTool As New ReportPrintTool(report)
			printTool.ShowPreviewDialog()
		End Sub

		Private Sub simpleButton4_Click(ByVal sender As Object, ByVal e As EventArgs) Handles simpleButton4.Click
			Dim report As New XtraReport1()

			Dim dataSource As XPObjectSource = CType(report.DataSource, XPObjectSource)
			AddHandler dataSource.ResolveSession, AddressOf OnResolveSession
			AddHandler dataSource.DismissSession, AddressOf OnDismissSession

			Dim printTool As New ReportPrintTool(report)
			printTool.ShowPreviewDialog()
		End Sub

		Private Sub OnResolveSession(ByVal sender As Object, ByVal e As ResolveSessionEventArgs)
			'Сreate a single instance of IDataLayer if it is not created
			If DataLayerInstance Is Nothing Then
				Dim connectionString As String = ConfigurationManager.ConnectionStrings("nwind").ConnectionString
				DataLayerInstance = XpoDefault.GetDataLayer(connectionString, DevExpress.Xpo.DB.AutoCreateOption.SchemaAlreadyExists)
			End If

			'Create new session based on the instance of DataLayer
			e.Session = New UnitOfWork(DataLayerInstance)
		End Sub

		Private Sub OnDismissSession(ByVal sender As Object, ByVal e As ResolveSessionEventArgs)
			e.Session.Session.Dispose()
		End Sub
	End Class
End Namespace
